ports/esp32/moducrypto.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014 Paul Sokolovsky
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "mbedtls/platform.h"
+#include "mbedtls/pk.h"
+#include "mbedtls/rsa.h"  // 20180505 MHE add rsa_sign
+
+//20180505 MHE------------------------------------------------------------------------------------------
+STATIC mp_obj_t mod_crypto_rsa_sign(mp_obj_t key, mp_obj_t message) {
+
+   
+    int ret;
+    mbedtls_pk_context pk;
+    mbedtls_rsa_context *rsa;
+    unsigned char hash[32];
+    mbedtls_pk_init( &pk );
+
+    // get the buffer to send from
+    mp_buffer_info_t bufkey;
+    mp_get_buffer_raise(key, &bufkey, MP_BUFFER_READ);
+    
+    // get the buffer to send from
+    mp_buffer_info_t bufmessage;
+    mp_get_buffer_raise(message, &bufmessage, MP_BUFFER_READ);
+   
+    ret = mbedtls_pk_parse_key(&pk, bufkey.buf, bufkey.len + 1, NULL, 0);
+    if ( ret != 0 ) {
+        mp_raise_ValueError("Unable to parse key");
+        goto exit;
+    }
+
+    if( !mbedtls_pk_can_do( &pk, MBEDTLS_PK_RSA ) )
+    {
+        mp_raise_ValueError("Not an RSA key");
+        goto exit;
+    }
+    
+    if( ( ret = mbedtls_md(
+                    mbedtls_md_info_from_type( MBEDTLS_MD_SHA256 ),
+                    bufmessage.buf, bufmessage.len, hash ) ) != 0 )
+    {
+        mp_raise_ValueError("Unable to hash message");
+        goto exit;

```

Comment:
If you call `mp_raise_ValueError(...)` then it throws an exception back to the caller and won't get to the `goto exit` line.  So this error handling needs to be refactored.